thrust/system/detail/generic/shuffle.inl

```diff
@@ -32,85 +32,98 @@ namespace system {
 namespace detail {
 namespace generic {
 
-// An implementation of RC5
-template <uint32_t num_rounds = 12> class rc5_bijection {
-private:
+// An implementation of a Feistel cipher for operating on 64 bit keys
+class feistel_bijection {
   struct round_state {
-    uint32_t A;
-    uint32_t B;
+    uint32_t left;
+    uint32_t right;
   };
 
-public:
+ public:
   template <class URBG>
-  __host__ __device__ rc5_bijection(uint64_t m, URBG &&g)
-   : w(get_cipher_bits(m))
-  {
-    init_state(std::forward<URBG>(g));
+  __host__ __device__ feistel_bijection(uint64_t m, URBG&& g) {
+    uint64_t total_bits = get_cipher_bits(m);
+    // Half bits rounded down
+    left_side_bits = total_bits / 2;
+    left_side_mask = (1ull << left_side_bits) - 1;
+    // Half the bits rounded up
+    right_side_bits = total_bits - left_side_bits;
+    right_side_mask = (1ull << right_side_bits) - 1;
+
+    for (uint64_t i = 0; i < num_rounds; i++) {
+      key[i] = g();
+    }
   }
 
-  __host__ __device__ uint64_t bijection_width() const {
-    return 1ull << (2*w);
+  __host__ __device__ uint64_t nearest_power_of_two() const {
+    return 1ull << (left_side_bits + right_side_bits);
   }
-
   __host__ __device__ uint64_t operator()(const uint64_t val) const {
-    if(w == 0)
-      return val;
-    round_state state = { (uint32_t)val & get_mask(), (uint32_t)(val >> w) };
-    state.A = (state.A + S[0]) & get_mask();
-    state.B = (state.B + S[1]) & get_mask();
-    for(uint32_t i = 0; i < num_rounds; i++)
+    // Extract the right and left sides of the input
+    uint32_t left = (uint32_t)(val >> right_side_bits);
+    uint32_t right = (uint32_t)(val & right_side_mask);
+    round_state state = {left, right};
+
+    for (uint64_t i = 0; i < num_rounds; i++) {
       state = do_round(state, i);
-    uint64_t res = state.B << w | state.A;
-    return res;
-  }
+    }
 
-private:
-  template <class URBG>
-  __host__ __device__ void init_state(URBG&& g)
-  {
-    thrust::uniform_int_distribution<uint32_t> dist(0, get_mask());
-    for( uint32_t i = 0; i < state_size; i++ )
-      S[i] = dist(g);
+    // Check we have the correct number of bits on each side
+    assert((state.left >> left_side_bits) == 0);
+    assert((state.right >> right_side_bits) == 0);
+
+    // Combine the left and right sides together to get result
+    return state.left << right_side_bits | state.right;
   }
 
+ private:
   // Find the nearest power of two
   __host__ __device__ uint64_t get_cipher_bits(uint64_t m) {
-    if(m == 0)
-      return 0;
     uint64_t i = 0;
-    m--;
     while (m != 0) {
       i++;
-      m >>= 2u;
+      m >>= 1;
```

Comment:
I believe there is a bug here where powers of 2 will be rounded up again.
We need to keep the guard for m == 0 and m--

Reply:
Good catch, any other fixes in your PR that I need to include?

Reply:
Not that I could see. I looked at the commit diff for the review and didn't notice anything